core/tee/tee_svc_storage.c

```diff
@@ -444,7 +445,8 @@ TEE_Result syscall_storage_obj_del(unsigned long obj)
 	if (o->pobj == NULL || o->pobj->obj_id == NULL)
 		return TEE_ERROR_BAD_STATE;
 
-	if (IS_ENABLED(CFG_NXP_SE05X)) {
+	/* Scan buffers that might contain a Secure Element key identifier */
+	if (IS_ENABLED(CFG_NXP_SE05X) && o->info.dataSize < SMALL_PAGE_SIZE) {
 		len = o->info.dataSize;
 		data = calloc(1, len);
```

Comment:
You should probably use `mempool_calloc()` and `mempool_free()` with `mempool_default` if can may need to alloc 4k. `mempool_calloc()` is more or less guaranteed to succeed on such small allocation.

Reply:
you are right. doing it next

Reply:
Actually this is a terrible idea leading to a deadlock on 6016. will leave it the way it is. 
This mempool thread dependent allocator is pure evil IMO

Reply:
Memory leakage will lead to deadlock, fast, but other than that I don't see how this could cause a deadlock.

Reply:
it is easy to reproduce - just run 6016 and it hits in a second or two

```
E/TC:? 3 ree_fs_read:368
E/TC:? 3 ree_fs_read_primitive:310
E/TC:? 3 get_pool:93            -- WAIT LOCK POOL
E/TC:? 3 get_pool:95            -- POOL LOCKED
E/TC:? 3 put_pool:114           -- POOL FINAL UNLOCK
E/TC:? 3 ree_fs_read_primitive:357
E/TC:? 3 ree_fs_read:370
E/TC:? 1 ree_fs_read:368
E/TC:? 1 ree_fs_read_primitive:310
E/TC:? 1 get_pool:93            -- WAIT LOCK POOL
E/TC:? 1 get_pool:95            -- POOL LOCKED
E/TC:? 1 put_pool:114           -- POOL FINAL UNLOCK
E/TC:? 1 ree_fs_read_primitive:357
E/TC:? 1 ree_fs_read:370
E/TC:? 2 get_pool:93            -- WAIT LOCK POOL
E/TC:? 2 get_pool:95            -- POOL LOCKED
E/TC:? 1 get_pool:93            -- WAIT LOCK POOL
E/TC:? 2 put_pool:114           -- POOL FINAL UNLOCK
E/TC:? 1 get_pool:95            -- POOL LOCKED
E/TC:? 2 read_dent:102 !!
E/TC:? 2 ree_fs_read_primitive:310
E/TC:? 2 get_pool:93            -- WAIT LOCK POOL

/* at this point the system locks up  */

```

Reply:
I can instrument it further if you think it is worth addressing the issue

Reply:
Please do, or if you can find a way of reproducing it on QEMU and I'll investigate.

I wonder what thread 1 is doing, waiting for some other mutex?

Reply:
the issue is with getting the mempool before taking the ree_fs_lock...is expected to be the other way around

Reply:
Thanks. The problem is with the locking order, sometimes `ree_fs_mutex` is locked first and sometimes `mempool_default->mu`.

It seems there's no easy way of using the mempool API at this stage.